### PR TITLE
Add in GetTickerCount, GetAndResetTickerCount, SetStatsLevel, and Get…

### PR DESCRIPTION
**Pull request by kjplatz**
This adds in Go hooks to access the Tickers within the rocksdb Options class (technically Statistics class within options).  The primary motivation for this is to evaluate the block cache hit/miss rates within RocksDB.   We can then use this information to make an informed decision on sizing the RocksDB cache versus the go fastcache we have sitting in front of some of our RocksDB databases.

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -69,6 +69,170 @@ const (
 	SkipAnyCorruptedRecordsRecovery      = WALRecoveryMode(3)
 )
 
+type Ticker uint32
+
+// Options for use in GetTickerCount and GetAndResetTickerCount
+// For definitions of individual flags, see <rocksdb/statistics.h>
+const (
```
Comment (akrylysov) [inline options.go:76]: These consts will have zero value. Take a look at [iota](https://github.com/golang/go/wiki/Iota) if you need every subsequent constant incremented.

File: gorocksdb.h [hunk 2]
```diff
@@ -28,3 +28,17 @@ extern void gorocksdb_mergeoperator_delete_value(void* state, const char* v, siz
 /* Slice Transform */
 
 extern rocksdb_slicetransform_t* gorocksdb_slicetransform_create(uintptr_t idx);
+
+/* Statistics/Tickers */
+#ifdef __cplusplus__
+extern "C" {
+#endif
```
Comment (akrylysov) [inline gorocksdb.h:35]: Should it follow same pattern as above for exporting functions?

File: options.go [hunk 3]
```diff
@@ -1211,3 +1375,25 @@ func (opts *Options) Destroy() {
...
+	return uint64(retval)
+}
+
+func (opts *Options) GetAndResetTickerCount(ticker Ticker) uint64 {
+	retval, _ := C.gorocksdb_get_and_reset_ticker_count(opts.c, C.uint(ticker))
+
+	return uint64(retval)
+}
+
+func (opts *Options) SetStatsLevel(sl StatsLevel) {
```
Comment (akrylysov) [inline options.go:1392]: Do we ever need to set stats?

File: options.go (hunk 1 above)
Comment (kjplatz) [reply options.go:76]: Changed to use iota.

File: gorocksdb.h (hunk 2 above)
Comment (kjplatz) [reply gorocksdb.h]: The extern "C" wasn't required until I added in C++ code.  I moved the extern "C" to the top of the file to make things more consistent.  Without this, cgo will fail to link in the functions in the C++ file.

File: options.go (hunk 3 above)
Comment (kjplatz) [reply options.go:1401]: We should never need to, but that's not what SetStatsLevel does.  I've documented SetStatsLevel with the appropriate information.

File: statistics.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+void rocksdb_set_stats_level(rocksdb_options_t *options, uint8_t level) {
+    options->rep.statistics->set_stats_level((rocksdb::StatsLevel)level);
+}
+
+uint8_t rocksdb_get_stats_level(rocksdb_options_t *options) {
+    return options->rep.statistics->get_stats_level();
+}
+
+void rocksdb_block_based_options_set_flush_every_key_policy(rocksdb_block_based_table_options_t *opts) {
```
Comment (akrylysov) [inline statistics.cpp:70]: Why is it a part of `statistics`?

File: options_block_based_table.go [hunk 5]
```diff
@@ -235,3 +235,9 @@ func (opts *BlockBasedTableOptions) SetFormatVersion(version int) {
 func (opts *BlockBasedTableOptions) SetIndexType(value IndexType) {
 	C.rocksdb_block_based_options_set_index_type(opts.c, C.int(value))
 }
+
+// SetFlushEveryBlockPolicy sets the table to perform a flush after inserting each key.  Useful for
```
Comment (akrylysov) [inline options_block_based_table.go:239]: Can we not add this function in this PR if it's only for testing? In tests you can call `Flush` if you need.

File: options.go [hunk 6]
```diff
@@ -1211,3 +1375,38 @@ func (opts *Options) Destroy() {
...
+
+// SetStatsLevel sets the current statistics logging level - See <rocksdb/statistics.h> for more information
+func (opts *Options) SetStatsLevel(sl StatsLevel) {
+	C.rocksdb_set_stats_level(opts.c, C.uchar(sl))
+}
+
+// GetStatsLevel retrieves the current statistics logging level - See <rocksdb/statistics.h> for more information
+func (opts *Options) GetStatsLevel() StatsLevel {
+	ret := C.rocksdb_get_stats_level(opts.c)
+    return StatsLevel(ret)
```
Comment (akrylysov) [inline options.go:1411]: Make sure the code is formatted with `go fmt` (I wish we had the CI setup for this repo).

File: statistics_test.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+import (
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"testing"
+)
+
+type stringMerge struct {
+}
+
+func (sm stringMerge) FullMerge(key, existingValue []byte, operands [][]byte) ([]byte, bool) {
```
Comment (akrylysov) [inline statistics_test.go:12]: Is there a reason to use `Merge` instead of simple `Put` for this test?

File: statistics_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	assert.True(it.Valid(), "Failed iterator seek")
+	assert.Equal("a3", string(it.Key().Data()), "Incorrect key found")
+	assert.Equal("x3", string(it.Value().Data()), "Incorrect value found")
+	assert.Equal(uint64(9), opts.GetTickerCount(BlockCacheMiss), "Incorrect block miss count")
+	assert.Equal(uint64(11), opts.GetTickerCount(BlockCacheHit), "Incorrect block hit count")
+
+	misses := opts.GetAndResetTickerCount(BlockCacheMiss)
+	hits := opts.GetAndResetTickerCount(BlockCacheHit)
+
+	assert.Equal(uint64(9), misses, "Incorrect block miss count")
```
Comment (akrylysov) [inline statistics_test.go:125]: Shouldn't hard-code these numbers. I can imagine the number of misses and hits depends on the implementation and may vary depending on RocksDB version.

File: statistics_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	t.Parallel()
+	assert := assert.New(t)
+	dir, err := ioutil.TempDir("/tmp", "gorocksdb-TestStatistics")
+	assert.NoError(err, "Could not create temp directory")
+
+	opts := NewDefaultOptions()
+	opts.SetCreateIfMissing(true)
+	mergeOpt := stringMerge{}
+	opts.SetMergeOperator(mergeOpt)
+	opts.EnableStatistics()
```
Comment (akrylysov) [inline statistics_test.go:46]: Where it's implemented?

**Pull request closed without merging**